### ISSUE TITLE
Saving model in TF 2.4

### ISSUE THREAD

**Issue by hoondy** [tensorflow/tensorflow#47479]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS 7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version: 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: cuda 11.1 cudnn 8.0.5
- GPU model and memory: A100, 40GB

**Describe the current behavior**

Saving a model in TF 2.4 omits sub-layers and generates warning messages about "untraced function."

**Describe the expected behavior**

No warning and the sub-layers will be directly callable after loading.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

EXAMPLE:
```
import tensorflow as tf
import numpy as np

print(tf.version.GIT_VERSION, tf.version.VERSION) ### v2.4.0-49-g85c8b2a817f 2.4.1

'''
simple autoencoder example
'''

class Encoder(tf.keras.layers.Layer):
    def __init__(self, name="encoder", **kwargs):
        super(Encoder, self).__init__(name=name, **kwargs)
        self.enc = tf.keras.layers.Dense(units=20, activation="relu", name='encoder_layer')

    def call(self, inputs):
        x = inputs
        x = self.enc(x)
        return x

class Decoder(tf.keras.layers.Layer):
    def __init__(self, name="decoder", **kwargs):
        super(Decoder, self).__init__(name=name, **kwargs)
        self.dec = tf.keras.layers.Dense(units=1000, activation="sigmoid", name='decoder_layer')

    def call(self, inputs):
        x = inputs
        x = self.dec(x)
        return x
    
class AutoEncoder(tf.keras.Model):
    def __init__(self, name="autoencoder", **kwargs):
        super(AutoEncoder, self).__init__(name=name, **kwargs)
        self.encoder = Encoder()
        self.decoder = Decoder()
    
    def call(self, inputs):
        x = self.encoder(inputs)
        x = self.decoder(x)
        return x

    def train_step(self, data):
        x, y = data
        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)
            mse_loss = tf.reduce_mean(tf.reduce_sum(tf.square(y - y_pred), axis=-1), axis=-1) # mean squared error
        gradients = tape.gradient(mse_loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
        return {"loss": mse_loss}
    
'''
example run
'''
inp = np.random.rand(10000,1000)
ae = AutoEncoder()
ae.compile(optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3))
ae.fit(inp, inp)

'''
saving throws warning message "Found untraced functions ... These functions will not be directly callable after loading."
This is new in TF 2.4.
TF 2.3 did not have this issue.
'''
ae.save('tmp')
```
OUTPUT:
```
v2.4.0-49-g85c8b2a817f 2.4.1
313/313 [==============================] - 1s 1ms/step - loss: 83.3821
WARNING:absl:Found untraced functions such as encoder_layer_layer_call_and_return_conditional_losses, encoder_layer_layer_call_fn, decoder_layer_layer_call_and_return_conditional_losses, decoder_layer_layer_call_fn, encoder_layer_layer_call_fn while saving (showing 5 of 10). These functions will not be directly callable after loading.
WARNING:absl:Found untraced functions such as encoder_layer_layer_call_and_return_conditional_losses, encoder_layer_layer_call_fn, decoder_layer_layer_call_and_return_conditional_losses, decoder_layer_layer_call_fn, encoder_layer_layer_call_fn while saving (showing 5 of 10). These functions will not be directly callable after loading.
INFO:tensorflow:Assets written to: tmp/assets
INFO:tensorflow:Assets written to: tmp/assets
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

Spawn from issue #44541

**Comment by amahendrakar** [comment]
@hoondy,
I was able to reproduce the issue with [TF v2.4.1](https://colab.research.google.com/gist/amahendrakar/57712f2e5ad9bca4a6cc7018e2b365f8/47479.ipynb) and TF-nightly. 

Whereas with [TF v2.3.2](https://colab.research.google.com/gist/amahendrakar/47503d1d9f3f40b93c66861f922906cb/47479-2-3.ipynb), I am facing a different warning stating `WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
.` 

Please check the linked gist for reference. Thanks!

**Comment by hoondy** [comment]
Additionally, I am seeing these new warning messages after training a slightly more complex model with larger data:

```
WARNING:tensorflow:AutoGraph could not transform <function canonicalize_signatures.<locals>.signature_wrapper at 0x2b9a0b138940> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Unable to locate the source code of <function canonicalize_signatures.<locals>.signature_wrapper at 0x2b9a0b138940>. Note that functions defined in certain environments, like the interactive Python shell do n
ot expose their source code. If that is the case, you should to define them in a .py source file. If you are certain the code is graph-compatible, wrap the call using @tf.autograph.do_not_convert. Original error: co
uld not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
```

**Comment by aks2193** [comment]
I am getting the same error with tf 2.4.1
Wanted to know is this just an issue with warning message or if I load the model from the saved checkpoint then will it cause issues with predictions

**Comment by ClementViricel** [comment]
Same warning with a simple one layer LSTM model with a lambda function. Please update the save function.

**Comment by NoelKennedy** [comment]
saving models in H5 format seems to work for me.  

`model.save(filepath, save_format="h5")`

**Comment by jvishnuvardhan** [comment]
@hoondy This is intended behavior and the warning is to encourage users to use `model.save` instead of tf.saved_model.save when saving keras models.

As mentioned in another similar issue, i trained with `TF2.4` and saved and then loaded with `tf-nightly`. I was able to train the `loaded_model` without any issue. The warning was only to encourage saving Keras model with `model.save` that saves layer configs / trainable status / name / other python attributes along with graph_def. 

Please let us know If you notice any error due to this warning. Thanks!

**Comment by hoondy** [comment]
Thanks for looking into this

**Comment by Luux** [comment]
@jvishnuvardhan as mentioned in https://github.com/tensorflow/tensorflow/issues/47554, these warnings occur when using model.save as well - not just tf.saved_model.save - and it seems to lead to other problems for some people. Maybe you could have a look?

**Comment by melioristic** [comment]
> @jvishnuvardhan as mentioned in #47554, these warnings occur when using model.save as well - not just tf.saved_model.save - and it seems to lead to other problems for some people. Maybe you could have a look?


I agree with this. The warning occurs with `model.save` as well, and then after loading the function if I train and save it again, I cannot load it. I can see this in **tensorlflow==2.6.0**

**Comment by aks2193** [comment]
Hey guys, so I was able to resolve the issue. 
I was using custom functions for loss and metric in my code and due to that was facing the issue. So included the get_config(self) function in those custom metrics code

And while loading from saved file added  below how it was loaded

saved_model = keras.models.load_model(
     f"{dsm.checkpoint_dir}/epoch_{EPOCHS}/model",
     custom_objects={
         "SparseTopKCategoricalAccuracyMetric": SparseTopKCategoricalAccuracyMetric,
         "custom_loss": custom_loss,
     },
 )
 
 Thanks everyone for all the help.
